### PR TITLE
[browser] Increase timeout and clean up after failed attempts of launching browser when testing

### PR DESCRIPTION
**Pull request by ilonatommy**
This is a playground for fixing https://github.com/dotnet/runtime/issues/107771.
The issue is hit only on Windows, where we are running WBT sequentially.

Changes:
- Upgrade Playwright to "1.47.0" (newest available)
- After upgrading Playwright, it is creating node dir in `linux-x64` dir, not in `linux`. Node requires elevated execution rights, so `chmod +x` path was changed accordingly.
- Killing all chrome processes before we call the problematic method that launches the browser.
- Making sure we clean up after each run by closing `Browser` and disposing of `Playwright`.

### CODE REVIEW

File: src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs [hunk 1]
```diff
@@ -128,17 +128,27 @@ public async Task<IBrowser> SpawnBrowserAsync(
...
             }
             catch (System.TimeoutException ex)
             {
                 attempt++;
                 _testOutput.WriteLine($"Attempt {attempt} failed with TimeoutException: {ex.Message}");
+
+                // Cleanup before retrying
+                if (Browser != null)
+                {
+                    await Browser.CloseAsync();
```
Comment (pavelsavara) [inline src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs:141]: wrap it with one more try-catch, in case Close or Dispose failed

**Review by pavelsavara — APPROVED** [review]

File: src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs [hunk 2]
```diff
@@ -128,17 +128,34 @@ public async Task<IBrowser> SpawnBrowserAsync(
...
+                    if (Playwright != null)
+                    {
+                        Playwright.Dispose();
+                        Playwright = null;
+                    }
+                }
+                catch(Exception ex)
+                {
+                    _testOutput.WriteLine($"Attempt to clean up failed with {ex}");
+                }
```
Comment (lewing) [inline src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs:155]:
```suggestion
                catch(Exception cleanupException)
                {
                    _testOutput.WriteLine($"Attempt to clean up failed with {cleanupException}");
                }
```

**Review by maraf — APPROVED** [review]

**Comment by ilonatommy** [comment]
It does not help:
```
        Attempt 3 failed with TimeoutException: System.TimeoutException: Timeout 15000ms exceeded.
=========================== logs ===========================
<launching> C:\helix\work\correlation\chrome-win\chrome.exe --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose,MediaRouter,AcceptCHFrame,AutoExpandDetailsElement,CertificateTransparencyComponentUpdater --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --export-tagged-pdf --headless --hide-scrollbars --mute-audio --blink-settings=primaryHoverType=2,availableHoverTypes=2,primaryPointerType=4,availablePointerTypes=4 --no-sandbox --explicitly-allowed-ports=49238 --ignore-certificate-errors --user-data-dir=C:\Users\ContainerAdministrator\AppData\Local\Temp\playwright_chromiumdev_profile-j7hftC --remote-debugging-pipe --no-startup-window
<launched> pid=9100
============================================================
           at Microsoft.Playwright.Transport.Connection.InnerSendMessageToServerAsync[T](String guid, String method, Object args) in /_/src/Playwright/Transport/Connection.cs:line 164
           at Microsoft.Playwright.Transport.Connection.WrapApiCallAsync[T](Func`1 action, Boolean isInternal) in /_/src/Playwright/Transport/Connection.cs:line 475
           at Microsoft.Playwright.Core.BrowserType.LaunchAsync(BrowserTypeLaunchOptions options) in /_/src/Playwright/Core/BrowserType.cs:line 61
           at Wasm.Build.Tests.BrowserRunner.SpawnBrowserAsync(String browserUrl, Boolean headless, Int32 timeout, Int32 maxRetries) in /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs:line 120
```
[log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-107865-merge-cc0b7029dc7646e7a5/Workloads-NoWebcil-ST-Wasm.Build.Tests.TestAppScenarios.AppSettingsTests/1/console.22f7b955.log?helixlogtype=result)

**Comment by lewing** [comment]
Can we check if there are other running chrome instances or something to help diagnose the issue?

**Comment by ilonatommy** [comment]
If we keep the last commit about killing all other chromes, running WBT locally can be a nuisance. If it works, we should check for running on CI and only then call `TerminateExistingChromeInstances`. I did not register local timeouts similar to these on CI, so it should be fine.

**Comment by ilonatommy** [comment]
It's not about chrome instances, `Found 0 old Chrome processes.` in [log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-107865-merge-975cec82dbc84acfa8/Workloads-NoWebcil-ST-Wasm.Build.Tests.TestAppScenarios.ModuleConfigTests/1/console.9c6b9260.log?helixlogtype=result) when it still timeouts.

**Comment by ilonatommy** [comment]
The chrome bump will be ready soon: https://github.com/dotnet/runtime/pull/107504, maybe that might help?

**Comment by ilonatommy** [comment]
Failures are on Windows with `Workloads-NoFingerprint-ST-Wasm.Build.Tests`:
```
 Wasm.Build.Tests.WasmTemplateTests.RunWithDifferentAppBundleLocations(forConsole: False, runOutsideProjectDirectory: True, extraProperties: "") [FAIL]
      System.Exception : Timed out after 120s waiting for 'WASM EXIT' message
      Stack Trace:
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(242,0): at Wasm.Build.Tests.BrowserRunner.WaitForExitMessageAsync(TimeSpan timeout)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(324,0): at Wasm.Build.Tests.WasmTemplateTests.BrowserRunTwiceWithAndThenWithoutBuildAsync(String config, String extraProperties, Boolean runOutsideProjectDirectory)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(325,0): at Wasm.Build.Tests.WasmTemplateTests.BrowserRunTwiceWithAndThenWithoutBuildAsync(String config, String extraProperties, Boolean runOutsideProjectDirectory)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(301,0): at Wasm.Build.Tests.WasmTemplateTests.RunWithDifferentAppBundleLocations(Boolean forConsole, Boolean runOutsideProjectDirectory, String extraProperties)
```

**Comment by ilonatommy** [comment]
Retry after the newest chrome bump

**Comment by ilonatommy** [comment]
> Can we check if there are other running chrome instances or something to help diagnose the issue?

Doesn't wbt on CI run in parallel? Here: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-107865-merge-5ed30c600b42473598/NoWorkload-NoWebcil-ST-Wasm.Build.Tests/1/console.c510d0c4.log?helixlogtype=result it looks as if we killed another test by cleaning up chrome processes while finishing another test.

**Comment by ilonatommy** [comment]
After log elevation it failed in a different way, on 
https://github.com/dotnet/runtime/blob/d22c74dc974a757b48380a01dd0450efba31db64/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs#L188
```
    Wasm.Build.Tests.WasmTemplateTests.RunWithDifferentAppBundleLocations(forConsole: False, runOutsideProjectDirectory: False, extraProperties: "") [FAIL]
      System.TimeoutException : Timeout 30000ms exceeded.
=========================== logs ===========================
navigating to "http://localhost:5281/", waiting until "load"
============================================================
      Stack Trace:
[] Executing (Captured Output) - C:\helix\work\workitem\e\dotnet-latest\dotnet.exe new wasmbrowser  -  in pwd C:\helix\work\workitem\e\wbt artifacts\browser_Release_p0w3rn4h_3di
        /_/src/Playwright/Transport/Connection.cs(164,0): at Microsoft.Playwright.Transport.Connection.InnerSendMessageToServerAsync[T](String guid, String method, Object args)
           at Microsoft.Playwright.Transport.Connection.WrapApiCallAsync[T](Func`1 action, Boolean isInternal)
        /_/src/Playwright/Core/Frame.cs(539,0): at Microsoft.Playwright.Core.Frame.GotoAsync(String url, FrameGotoOptions options)
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(232,0): at Wasm.Build.Tests.BrowserRunner.RunAsync(IBrowserContext context, String browserUrl, Boolean headless, Action`2 onConsoleMessage, Action`1 onError, Func`2 modifyBrowserUrl, Boolean resetExitedState)
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(201,0): at Wasm.Build.Tests.BrowserRunner.RunAsync(ToolCommand cmd, String args, Boolean headless, Action`2 onConsoleMessage, Action`1 onServerMessage, Action`1 onError, Func`2 modifyBrowserUrl)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(323,0): at Wasm.Build.Tests.WasmTemplateTests.BrowserRunTwiceWithAndThenWithoutBuildAsync(String config, String extraProperties, Boolean runOutsideProjectDirectory)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(325,0): at Wasm.Build.Tests.WasmTemplateTests.BrowserRunTwiceWithAndThenWithoutBuildAsync(String config, String extraProperties, Boolean runOutsideProjectDirectory)
        /_/src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs(301,0): at Wasm.Build.Tests.WasmTemplateTests.RunWithDifferentAppBundleLocations(Boolean forConsole, Boolean runOutsideProjectDirectory, String extraProperties)
        --- End of stack trace from previous location ---
```
Rerunning to see if that's a rule.

**Comment by ilonatommy** [comment]
Log:
```
System.ComponentModel.Win32Exception : An error occurred trying to start process '/root/helix/work/workitem/e/.playwright/node/linux-x64/node' with working directory '/root/helix/work/workitem/e'. Permission denied
      Stack Trace:
           at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
           at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
           at System.Diagnostics.Process.Start()
        /_/src/Playwright/Transport/StdIOTransport.cs(168,0): at Microsoft.Playwright.Transport.StdIOTransport.StartProcessWithUTF8IOEncoding(Process process)
        /_/src/Playwright/Transport/StdIOTransport.cs(47,0): at Microsoft.Playwright.Transport.StdIOTransport..ctor()
        /_/src/Playwright/Playwright.cs(43,0): at Microsoft.Playwright.Playwright.CreateAsync()
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(128,0): at Wasm.Build.Tests.BrowserRunner.SpawnBrowserAsync(String browserUrl, Boolean headless, Int32 timeout, Int32 maxRetries, String language)
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(201,0): at Wasm.Build.Tests.BrowserRunner.RunAsync(ToolCommand cmd, String args, Boolean headless, String language, Action`2 onConsoleMessage, Action`1 onServerMessage, Action`1 onError, Func`2 modifyBrowserUrl)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(202,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(229,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/WorkloadRequiredTests.cs(119,0): at Wasm.Build.Tests.Blazor.WorkloadRequiredTests.WorkloadNotRequiredForInvariantGlobalization(String config, Boolean invariant, Boolean publish)
        --- End of stack trace from previous location ---
```
It's a problem only on linux, so I will add a task to yml that adds exec permission to node.

File: src/mono/wasm/Wasm.Build.Tests/Wasm.Build.Tests.csproj [hunk 3]
```diff
@@ -81,6 +81,8 @@
       <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="export CHROME_PATH_FOR_TESTS=$HELIX_CORRELATION_PAYLOAD/$(ChromeDirName)/$(ChromeBinaryName)" />
       <RunScriptCommands Condition="'$(OS)' == 'Windows_NT'" Include="set CHROME_PATH_FOR_TESTS=%HELIX_CORRELATION_PAYLOAD%\$(ChromeDirName)\$(ChromeBinaryName)" />
 
+      <!-- <RunScriptCommands Condition="'$(OS)' == 'Windows_NT'" Include="set DEBUG=pw:browser*" /> -->
```
Comment (ilonatommy) [inline src/mono/wasm/Wasm.Build.Tests/Wasm.Build.Tests.csproj:84]: ToDo: remove when finished.

File: src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs [hunk 4]
```diff
@@ -118,6 +120,12 @@ public async Task<IBrowser> SpawnBrowserAsync(
         {
             try
             {
+                // widnows runs are sequential, we cannot terminate when we run in parallel
+                if (s_runsOnWindows)
+                {
+                    TerminateExistingChromeInstances();
```
Comment (pavelsavara) [inline src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs:126]: should we also terminate after each test ?
do we believe that chrome process survived on the agent from another run ? this is docker/VM right ?

Comment (ilonatommy) [reply src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs:126]: I don't believe after is needed. The problem is with `LaunchAsync`, so the idea was to make sure nothing is running directly before calling that method. And yes, the suspicion was that some leftovers from the previous test was left and that's why new browser launch is time outing. 
This failure is only on CI, I have never managed to reproduce it locally.

**Comment by ilonatommy** [comment]
> [Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-107865-merge-868f283d28724a16b6/NoWorkload-NoWebcil-ST-Wasm.Build.Tests/1/console.f06fbac0.log?helixlogtype=result):
> 
> ```
> System.ComponentModel.Win32Exception : An error occurred trying to start process '/root/helix/work/workitem/e/.playwright/node/linux-x64/node' with working directory '/root/helix/work/workitem/e'. Permission denied
>       Stack Trace:
>            at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
>            at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
>            at System.Diagnostics.Process.Start()
>         /_/src/Playwright/Transport/StdIOTransport.cs(168,0): at Microsoft.Playwright.Transport.StdIOTransport.StartProcessWithUTF8IOEncoding(Process process)
>         /_/src/Playwright/Transport/StdIOTransport.cs(47,0): at Microsoft.Playwright.Transport.StdIOTransport..ctor()
>         /_/src/Playwright/Playwright.cs(43,0): at Microsoft.Playwright.Playwright.CreateAsync()
>         /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(128,0): at Wasm.Build.Tests.BrowserRunner.SpawnBrowserAsync(String browserUrl, Boolean headless, Int32 timeout, Int32 maxRetries, String language)
>         /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(201,0): at Wasm.Build.Tests.BrowserRunner.RunAsync(ToolCommand cmd, String args, Boolean headless, String language, Action`2 onConsoleMessage, Action`1 onServerMessage, Action`1 onError, Func`2 modifyBrowserUrl)
>         /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(202,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
>         /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(229,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
>         /_/src/mono/wasm/Wasm.Build.Tests/Blazor/WorkloadRequiredTests.cs(119,0): at Wasm.Build.Tests.Blazor.WorkloadRequiredTests.WorkloadNotRequiredForInvariantGlobalization(String config, Boolean invariant, Boolean publish)
>         --- End of stack trace from previous location ---
> ```
> 
> It's a problem only on linux, so I will add a task to yml that adds exec permission to node.

@akoeplinger, do you know why it did not help? It's still failing with same message

**Comment by ilonatommy** [comment]
The issue is still existing after the Playwright bump, see: https://helix.dot.net/api/2019-06-17/jobs/8d53b01d-113d-4e65-a893-6a5d825a52b6/workitems/Workloads-NoWebcil-ST-Wasm.Build.Tests.TestAppScenarios.LibraryInitializerTests/console.
I created an issue in Playwright repo: https://github.com/microsoft/playwright/issues/32897, maybe we can get some suggestions about what could be going on.

**Comment by ilonatommy** [comment]
Using default chromium from Playwright is broken on Linux:
```
Microsoft.Playwright.PlaywrightException : Executable doesn't exist at /home/helixbot/.cache/ms-playwright/chromium-1134/chrome-linux/chrome
      ╔════════════════════════════════════════════════════════════╗
      ║ Looks like Playwright was just installed or updated.       ║
      ║ Please run the following command to download new browsers: ║
      ║                                                            ║
      ║     pwsh bin/Debug/netX/playwright.ps1 install             ║
      ║                                                            ║
      ║ <3 Playwright Team                                         ║
      ╚════════════════════════════════════════════════════════════╝
      Stack Trace:
        /_/src/Playwright/Transport/Connection.cs(206,0): at Microsoft.Playwright.Transport.Connection.InnerSendMessageToServerAsync[T](ChannelOwner object, String method, Dictionary`2 dictionary, Boolean keepNulls)
        /_/src/Playwright/Transport/Connection.cs(532,0): at Microsoft.Playwright.Transport.Connection.WrapApiCallAsync[T](Func`1 action, Boolean isInternal)
        /_/src/Playwright/Core/BrowserType.cs(56,0): at Microsoft.Playwright.Core.BrowserType.LaunchAsync(BrowserTypeLaunchOptions options)
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(129,0): at Wasm.Build.Tests.BrowserRunner.SpawnBrowserAsync(String browserUrl, Boolean headless, Nullable`1 timeout, Int32 maxRetries, String language)
        /_/src/mono/wasm/Wasm.Build.Tests/BrowserRunner.cs(201,0): at Wasm.Build.Tests.BrowserRunner.RunAsync(ToolCommand cmd, String args, Boolean headless, String language, Action`2 onConsoleMessage, Action`1 onServerMessage, Action`1 onError, Func`2 modifyBrowserUrl)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(202,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/BlazorWasmTestBase.cs(229,0): at Wasm.Build.Tests.BlazorWasmTestBase.BlazorRunTest(String runArgs, String workingDirectory, BlazorRunOptions runOptions)
        /_/src/mono/wasm/Wasm.Build.Tests/Blazor/WorkloadRequiredTests.cs(119,0): at Wasm.Build.Tests.Blazor.WorkloadRequiredTests.WorkloadNotRequiredForInvariantGlobalization(String config, Boolean invariant, Boolean publish)
        --- End of stack trace from previous location ---
```
https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-107865-merge-50c72d5eddc049108b/NoWorkload-NoWebcil-ST-Wasm.Build.Tests/1/console.47c37d30.log?helixlogtype=result

**Comment by ilonatommy** [comment]
1st run with default chrome + higher timeout: success.
Rerunning.

**Pull request closed without merging**